feat: unify settings into single EditSettings page

Sekhmet opened the pull request:
### Summary

This PR merges Settings and EditSettings into single page.
It also adds reusable `ModalTransactionProgress` modal that you can use to track actions (can be also used for Voting progress).

Closes: https://github.com/snapshot-labs/workflow/issues/165

[Designs on Figma](https://www.figma.com/design/hmHTau1Y1KYzrBBKLQr5QV/%23design?node-id=5248-3827&t=IGnSKZSczvUiHerC-0)

### How to test

1. Go to space you are controller of (onchain only).
2. You see Settings tab and you can open it.
3. You can see different settings (strategies, voting, executions, controller).
4. You can modify those settings and changes are detected (there is some delay because metadata needs to be created async).
5. You can reset changes.
6. You can modify settings and go through new ModalTransactionProgress flow (you can reject in the middle of it, wait for confirmation etc.
7. You can modify controller and go through new ModalTransactionProgress flow.

### Screenshots


https://github.com/user-attachments/assets/5aac8fd8-aeb6-43a7-a925-443b86b685d2

bonustrack commented:
Logo is missing on "create" page
<img width="1467" alt="image" src="https://github.com/user-attachments/assets/da593c31-6fff-4fd6-b5c0-8498ee0be392">

ChaituVR commented on apps/ui/src/router.ts:
```diff
@@ -43,7 +42,6 @@ const routes: any[] = [
       { path: '', name: 'space-overview', component: SpaceOverview },
       { path: 'proposals', name: 'space-proposals', component: SpaceProposals },
       { path: 'search', name: 'space-search', component: SpaceSearch },
-      { path: 'settings', name: 'space-settings', component: SpaceSettings },
```
There won't be a way for others to view settings now.? it was easy to check/debug space settings

Sekhmet replied:
Yeah, that is one concern I have as well, but this wasn’t publicly visible (but helpful for debugging). I wanted to make it visible for others, but we didn’t do it because it was supposed to change with new Edit Setting page, but looks like it’s still issue as now normal users can’t see it, and we won’t be able to use it for debugging as well (unless we pretend to be the controller).

@bonustrack

bonustrack replied:
This page should be visible for anyone (even unlogged) just not accessible from the sidebar if you are not logged with the controller address. I would also use "/settings" instead of "/edit-settings" route for that page.

Sekhmet replied:
Right now this page just displays a message if you are not controller, what should we do if non-controller opens it?

Should we just block save actions? Or have special read-only mode (because there you can add/edit/remove strategies which would be accessible to non-controller), I guess if it's hidden by default just disabling save action might be enough.

Sekhmet commented:
@bonustrack can you take a look into if UI text is what we want in this? There are some placeholder texts as well as there were few on designs as well.

bonustrack commented:
@Sekhmet Yes it's good

Here are the description for controller:
```
The controller is the account able to change the space settings and cancel pending proposals.
```

Here is for voting:
```
Set the proposal delay, minimum duration, which is the shortest time needed to execute a proposal if quorum passes, and maximum duration for voting.
```

Here is for execution:
```
Execution strategies determine if a proposal passes and how it is executed. This section is currently read-only.
```

Also "Proposal threshold" should not appear in "Voting" section

Sekhmet commented:
> There is some kind of memory leak issue that freeze the UI on the Nimbora space:
https://snapshot.box/#/sn:0x05ea5ef0c54c84dc7382629684c6e536c0b06246b3b0981c426b42372e3ef263

This was caused by Whitelist strategy, because it's merkle tree generating it proof is taking some time (and it's synchronous) and this is what final param for strategy is, added few changes that will now prevent computing `params` for MerkleWhitelist strategies and instead compare metadata only (which are equivalent because it contains whitelist itself).

I also had to refactor how we detect changes in validation strategy, 

> Also the active strategy items are not vertically aligned.

Can we update it on Figma? Current alignment is result of line heights of the rows we have so if I change it it will affect space between those.

Also you were checking paddings on top and bottom text, but there is also icon that affects alignment. Not sure if we should have equal padding relative to text or icon.

bonustrack reviewed (approved):
tACK, we should get update on Figma soon

ChaituVR commented on apps/ui/src/components/FormValidation.vue:
```diff
@@ -48,7 +50,7 @@ function handleStrategySave(value: Record<string, any>) {
 }
 
-onMounted(() => {
+watchEffect(() => {
```
Maybe better to use `computed`? instead of `watchEffect` and `ref`

```ts
const votingStrategies = computed(() => model.value?.params?.strategies || []);
```

Sekhmet replied:
Had to use computed with get/set for it, because we also write to it.

ChaituVR commented on apps/ui/src/views/Space/Settings.vue:
I think not related to current changes but getting this notice whenever I open the proposal validation modal

![image](https://github.com/user-attachments/assets/c63f21a2-0128-4bac-89a0-f4fc30cded9f)

Sekhmet replied:
Yeah, unrelated to this PR but fixed this issue.

ChaituVR commented on apps/ui/src/components/Modal/ChangeController.vue:
```diff
@@ -0,0 +1,115 @@
+
+const props = defineProps<{
+  open: boolean;
+  initialState?: any;
```
```suggestion
  initialState?: { controller: string };
```

ChaituVR commented on apps/ui/src/components/Modal/ChangeController.vue:
```diff
@@ -0,0 +1,115 @@
+      <UiButton
+        class="w-full"
+        :disabled="Object.keys(formErrors).length > 0"
+        @click="emit('save', form.controller)"
```
sometimes we are using `$emit` and sometimes we are using `emit` 🤔 i am not too sure what's the difference

wa0x6e replied:
I think we should always use `$emit` inside `<template>`, and `emit` inside `<script>`.

I believe we got a similar debate about that with Sam some time ago

Sekhmet replied:
The only difference is that `$emit` is special keyword that you can use in Vue templates out of the box, while `emit` can be only used if you assign it in script (it could actually be called anything).
```ts
const emit = defineEmits(...)
// you can name it whatever you want and use it in templates via that name
```

I think I usually end up using named emit vs special keyword, but in most cases I end up using the same format that the file I copied source from 😆 

I don't see any eslint rules to enforce it.

wa0x6e replied:
I remember that we've already got confused once https://github.com/snapshot-labs/sx-monorepo/pull/100#discussion_r1508844944

Maybe just always assign the emit to a constant, and avoid the magical `$emit` from now ?

Sekhmet replied:
> Maybe just always assign the emit to a constant, and avoid the magical $emit from now ?

That's what I'd prefer, but we can discuss it and later on convert the codebase. Might be annoying to enforce without eslint rule for it.

ChaituVR commented on apps/ui/src/components/Modal/TransactionProgress.vue:
```diff
@@ -0,0 +1,186 @@
+    emit('confirmed');
+    step.value = 'success';
+  } catch (e) {
+    console.warn(e);
```
```suggestion
    console.warn('Transaction failed:', e);
```

ChaituVR commented on apps/ui/src/components/Modal/TransactionProgress.vue:
```diff
@@ -0,0 +1,186 @@
+  throw new Error('Invalid step');
+});
+
+function sleep(ms: number) {
+  return new Promise(resolve => setTimeout(resolve, ms));
+}
```
can go to utils? could be helpful in other places in future

ChaituVR reviewed (approved):
tAck

wa0x6e commented on apps/ui/src/composables/useActions.ts:
```diff
@@ -125,6 +125,8 @@ export function useActions() {
       );
     }
+
+    return envelope.trasacton_hash || envelope.hash;
```
```suggestion
    return envelope.transaction_hash || envelope.hash;
```

wa0x6e replied:
Not sure this is always valid, as `envelope` does not always have `hash` and `transaction_hash` properties.

Looking at line 108, we need to fallback to using the hash inside `receipt` instead.

wa0x6e replied:
On my vote modal PR, I used this fix https://github.com/snapshot-labs/sx-monorepo/pull/627/files#diff-5daf95b187b0b02b358aef78e66671695bdf3e693fd7a9bc9d56068e27bf056e

wa0x6e commented:
The transaction loading icon is not dark theme friendly, need to reverse the color 

![Screenshot 2024-08-14 at 01 41 09](https://github.com/user-attachments/assets/47b75703-ac9d-4291-b797-4b585fb23cd3)

bonustrack commented:
Instead of this SVG can we use our default loading icon animation and just resize it?

Sekhmet commented:
> Instead of this SVG can we use our default loading icon animation and just resize it?

Updated but I don't think it looks too good.

bonustrack commented:
I think it's fine, I'm only bothered by the speed of the animation that seem a bit too fast, if you can increase it to 1sec that should work.

bonustrack commented:
![sample](https://github.com/user-attachments/assets/f6ca4fe6-13ce-47c7-89af-26736f83bd3d)

wa0x6e commented:
I only tester the transaction modal so far, but should be good, since the other tAcked it.

I know the edit page is only accessible by a hidden direct link for now, but maybe should handle the case where someone open this page on an offchain space.

wa0x6e commented:
Another little issue, in the voting delay tab

- Edit the voting delay: change the value by something else, then accept the change
- The new value is reflected in the display
- Click on the "Reset" button
- In the display, the voting delay should be back to old value
- Open the voting delay form, still pre-filled with edited new value

wa0x6e commented on apps/ui/src/views/Space/Settings.vue:
```diff
@@ -36,271 +166,533 @@ function formatCurrentValue(value: number) {
+
+<template>
+  <div
+    class="overflow-y-scroll no-scrollbar z-40 sticky top-[71px] lg:top-[72px]"
```
Shouldn't this be `overflow-x-scroll`

wa0x6e commented:
And in the VOTING tab, all form should exit the edit mode on reset

wa0x6e commented:
On narrow screen, the intercom button is overlapping the save button

![Screenshot 2024-08-14 at 12 33 06](https://github.com/user-attachments/assets/40d0ddf9-67c4-4746-be96-ad8c7ab2abbf)

wa0x6e commented:
On some small screen, the bottom toolbar layout is messed up.

![Screenshot 2024-08-14 at 12 35 33](https://github.com/user-attachments/assets/d73a39a0-d085-4bb3-b559-595773711ac7)

Maybe add a breakpoint, and have the message on one line, and the buttons on a second line

wa0x6e commented:
When saving a space settings without proposal validation, it return

![Screenshot 2024-08-14 at 12 50 21](https://github.com/user-attachments/assets/61708b9a-66d6-4c42-bd2a-d3dbf2ac3a4a)

In the old edit-settings page, the "Save" button is disabled if the form is invalid (but still not showing any validation error)

Sekhmet commented:
Addresses all comments except for
> On narrow screen, the intercom button is overlapping the save button

Not sure what to do with it, to me it's just super annoying in all cases and will always interrupt with our UI.

bonustrack commented:
The Figma been updated with proper horizontal alignment for cards

wa0x6e commented:
Do you think your scroller horizontal can be merged with #648 ?

Sekhmet commented:
@wa0x6e I think it should be possible to merge those two. We will need to add some props like `enableButtons` to add interactions via buttons and `showOnlyWheenNeeded` to handle this usecase (here gradient is only visible if you are not scrolled enough.

I guess we can merge your PR first and then I will see how to adapt it for our usecase.
